lib/room.rb

```diff
@@ -0,0 +1,85 @@
+class Room
+  attr_reader :name, :css_class, :presence_colour_rgb
+
+  def initialize(name:, css_class:, gcal_identifier:, presence_colour_rgb:)
+    @name = name
+    @css_class = css_class
+    @gcal_identifier = gcal_identifier
+    @presence_colour_rgb = presence_colour_rgb
+    @events_cache_expires = Time.now
+  end
+
+  def empty
+    events.select{ |event| event[:now] }.empty?
+  end
+
+  def upcoming_event_today
+    events.select{ |event| not event[:now] }.any?
+  end
+
+  def empty_until_string
+    if upcoming_event_today
+      events[0][:start_time_string]
+    else
+      "Tomorrow"
+    end
+  end
+
+  def minutes_to_next_event
+    if events.empty?
+      false
+    else
+      if empty
+        ((events[0][:start_time] - DateTime.now) * 24 * 60).to_i
+      else
+        if upcoming_event_today
+          ((events[1][:start_time] - DateTime.now) * 24 * 60).to_i
+        else
+          false
+        end
+      end
+    end
+  end
+
+  def minutes_to_end_of_event
+    if events.empty?
+      false
+    else
+      ((events[0][:end_time] - DateTime.now) * 24 * 60).to_i
+    end
+  end
+
+  def events
+    if (@events_cache_expires < Time.now)
+      @cached_events = fetch_events(@gcal_identifier).map{ |event| {
+          :summary => event.summary || 'Private or unspecified',
+          :start_time => event.start.date_time,
+          :start_time_string => event.start.date || event.start.date_time.strftime("%l:%M %P"),
+          :end_time => event.end.date_time,
+          :end_time_string => event.end.date || event.end.date_time.strftime("%l:%M %P"),
+          :organiser => event.organizer ? ( event.organizer.display_name || event.organizer.email) : 'Private or unspecified',
+          :now => DateTime.now.between?(event.start.date_time, event.end.date_time)
+        }
+      }
+      @events_cache_expires = Time.now + CACHE_EXPIRY_TIMEOUT
+    end
+
+    @cached_events
+  end
+
+  # Fetch the next 5 events today for this room
+  def fetch_events(calendar_id)
+    response = service.list_events(calendar_id,
```

Comment:
This is now the only place that `service` is being invoked.

Would it make sense to `require` it only in the `room` file, or is this just splitting hairs?

Reply:
Personal preference is to do imports where they're needed, even if it's a double import. But then I'm accustomed to modern JavaScript where nothing leaks out of file scope.

Reply:
It would definitely be cleaner if it was only imported in `room.rb`, otherwise someone in the future may replace this and we'd still be left with a stray import kicking around.